prusti-viper/src/encoder/vir/program.rs

```diff
@@ -0,0 +1,94 @@
+use super::*;
+use prusti_interface::config;
+use viper::{self, AstFactory};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Program {
+    pub fields: Vec<Field>,
+    pub builtin_methods: Vec<BodylessMethod>,
+    pub methods: Vec<CfgMethod>,
+    pub functions: Vec<Function>,
+    pub viper_predicates: Vec<Predicate>,
+}
+
+impl Program {
+    pub fn optimized(mut self) -> Self {
+        // can't borrow self because we need to move fields
+        let (new_methods, new_functions) =
+            optimizations::functions::inline_constant_functions(self.methods, self.functions);
+        self.methods = new_methods
+            .into_iter()
+            .map(|m| {
+                let purified = optimizations::methods::purify_vars(m);
+                optimizations::folding::FoldingOptimizer::optimize(purified)
+            })
+            .collect();
+        self.functions = new_functions
+            .into_iter()
+            .map(|f| optimizations::folding::FoldingOptimizer::optimize(f))
+            .collect();
+        self
+    }
+
+    pub fn to_viper<'v>(self, ast: &AstFactory<'v>) -> viper::Program<'v> {
```

Comment:
Could you move this method to the `to_viper` module, implementing the `ToViper` trait for `Program`?

Reply:
Oh yeah, for some reason I got confused awhile back when this idea first came up. Because the code uses `into_iter()`, which requires ownership, I thought I couldn't easily implement `ToViper`, but it turns out it works just fine with regular old `iter()`, so I've implemented the trait now.